core/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTest.java

```diff
@@ -0,0 +1,154 @@
+        if (!queryBuilder.fieldName().equals(STRING_FIELD_NAME)) {
+            return;
+        }
+
+        // expected returned terms depending on whether we have a terms query or a terms lookup query
+        List<Object> terms;
+        if (queryBuilder.termsLookup() != null) {
+            TermsLookup termsLookup = queryBuilder.termsLookup();
+            terms = Lists.newArrayList((Object) termsLookup.getIndex(), termsLookup.getType(),
+                    termsLookup.getId(), termsLookup.getRouting(), termsLookup.getPath());
+        } else {
+            terms = queryBuilder.values();
+        }
+
+        // compare whether we have the expected list of terms returned
+        Iterator<Object> iter = terms.iterator();
+        for (BooleanClause booleanClause : booleanQuery) {
+            assertThat(booleanClause.getQuery(), instanceOf(TermQuery.class));
+            Term term = ((TermQuery) booleanClause.getQuery()).getTerm();
+            Object next = iter.next();
+            if (next == null) {
+                continue;
+            }
+            assertThat(term, equalTo(new Term(queryBuilder.fieldName(), next.toString())));
+        }
+    }
+
+    @Test
+    public void testValidate() {
+        TermsQueryBuilder termsQueryBuilder = new TermsQueryBuilder(null, "term");
+        assertThat(termsQueryBuilder.validate().validationErrors().size(), is(1));
+
+        termsQueryBuilder = new TermsQueryBuilder("field", "term").termsLookup(randomTermsLookup());
+        assertThat(termsQueryBuilder.validate().validationErrors().size(), is(1));
+
+        termsQueryBuilder = new TermsQueryBuilder(null, "term").termsLookup(randomTermsLookup());
+        assertThat(termsQueryBuilder.validate().validationErrors().size(), is(2));
+
+        termsQueryBuilder = new TermsQueryBuilder("field", "term");
+        assertNull(termsQueryBuilder.validate());
+    }
+
+    @Test
+    public void testValidateLookupQuery() {
+        TermsQueryBuilder termsQuery = new TermsQueryBuilder("field").termsLookup(new TermsLookup());
+        int totalExpectedErrors = 3;
+        if (randomBoolean()) {
+            termsQuery.lookupId("id");
+            totalExpectedErrors--;
+        }
+        if (randomBoolean()) {
+            termsQuery.lookupType("type");
+            totalExpectedErrors--;
+        }
+        if (randomBoolean()) {
+            termsQuery.lookupPath("path");
+            totalExpectedErrors--;
+        }
+        assertValidate(termsQuery, totalExpectedErrors);
+    }
+
+    @Test
+    public void testNullValues() {
+        QueryShardContext context = createShardContext();
+        context.setAllowUnmappedFields(true);
+        try {
+            new TermsQueryBuilder("field", (Iterable) null);
+            fail("should have failed with IllegalArgumentException");
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), Matchers.containsString("No value specified for terms query"));
+        }
+    }
```

Comment:
can we add a test that verifies that if you specify both terms_lookup and values on the REST layer we get back a validation error? This part in the parser seems a bit tricky, I want to make sure we don't lose this.